default/python/universe_generation/monsters.py

```diff
@@ -115,62 +118,60 @@ def generate_monsters(monster_freq, systems):
     tracked_plan_tries = {name: 0 for name in nest_name_map.values()}
     tracked_plan_counts = {name: 0 for name in nest_name_map.values()}
     tracked_plan_valid_locations = {fp: 0 for fp in fleet_plans if fp.name() in tracked_plan_counts}
-    tracked_nest_valid_locations = {nest: 0 for nest in nest_name_map}
 
     if not fleet_plans:
         return
 
     universe = fo.get_universe()
 
+    EMPIRE_TO_MONSTER_MIN_DIST = 2
+    empire_exclusions = set(itertools.chain.from_iterable(
+        get_systems_within_jumps(e, EMPIRE_TO_MONSTER_MIN_DIST)
+        for e in home_systems))
+    systems = [s for s in _systems if s not in empire_exclusions]
```

Comment:
I understand the desire to speed things up, but you can't do that here (or at least, it's not a good idea to do that here at all). Unlike natives, monsters do have a location condition (yes, that's an inconsistency!), and all criteria that determine where a monster can be spawned should be defined there.

That part here effectively adds another location restriction (to the ones defined in a monsters location condition), which means the rules where a monster can be placed becomes scattered among very different files/locations. To give an example how this can screw up what a content scripter might want to do:

Imagine someone comes up with the idea to script some kind of "invisible helper monster" that can appear close to player starting homeworlds, and that secretly aids the player (by applying some boosts or whatever). The above code would prevent that monster from ever spawning, with no hint whatsoever for the content scripter to figure out why their script doesn't work as intended.

The point being, you can't make *any* assumptions about space monsters and where they should/shouldn't spawn here. Space monsters can and are used for all kind of things, not just the obvious nasty critters that are supposed to make life hard for players. 😉 

By the way an excellent example why I want to strip those location conditions from monsters and specials and integrate them into the universe generation scripts. Among other things, to get rid of inconsistencies, to allow for optimizations like you want to do here without making life hard for content scripters, etc.